src/aproxy.py

```diff
@@ -377,12 +386,25 @@ def check_relation_availability(self) -> None:
         """Check if the Juju relation is available for topology resolution.
 
         Raises:
-            TopologyUnavailableError: If the relation or binding information is unavailable.
+            RelationMissingError: If the required relation is missing.
+            TopologyUnavailableError: If the binding information is unavailable.
         """
-        try:
-            self._get_primary_ip()
-        except TopologyUnavailableError as e:
-            raise TopologyUnavailableError(f"Failed to get unit IP: {e}") from e
+        relation = self.charm.model.get_relation(RELATION_NAME)
+        binding = self.charm.model.get_binding(RELATION_NAME)
+
+        if not relation:
+            raise RelationMissingError(
+                f"Missing relation '{RELATION_NAME}' to the principal charm."
+            )
+
+        if not binding or binding.network is None:
+            raise TopologyUnavailableError(
+                f"Relation '{RELATION_NAME}' network not available when trying to get unit IP."
+            )
+
+        # Set for later use to get primary IP
+        self.relation = relation
+        self.binding = binding
```

Comment:
That's a side effect of this method, which then takes the assumption that this method was called before, I'd rather call this method multiple times than having such a side effect